### PR TITLE
VB-842 Calculate the cost of our service

### PR DESCRIPTION
**Pull request by AledWEvans**
VB-842 Calculate the cost of our service

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/controller/VisitController.kt [hunk 1]
```diff
@@ -223,8 +225,10 @@ class VisitController(
     @Schema(description = "applicationReference", example = "dfs-wjs-eqr", required = true)
     @PathVariable
     applicationReference: String,
+    @RequestBody @Valid
+    bookingRequestDto: BookingRequestDto,
   ): VisitDto {
-    return visitService.bookVisit(applicationReference.trim())
+    return visitService.bookVisit(applicationReference.trim(), bookingRequestDto)
```
Comment (dhirajshettymoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/controller/VisitController.kt:231]: will these details not be the same as the reserved visit?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/repository/EventAuditRepository.kt [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+import uk.gov.justice.digital.hmpps.visitscheduler.model.ApplicationMethodType
+import uk.gov.justice.digital.hmpps.visitscheduler.model.entity.EventAudit
+
+@Repository
+interface EventAuditRepository : JpaRepository<EventAudit, Long> {
+  fun findByBookingReference(bookingReference: String): List<EventAudit>
+
+  @Query(
+    "SELECT * FROM event_audit " +
+      "WHERE booking_reference = :bookingReference AND type = 'BOOKED_VISIT'  " +
```
Comment (dhirajshettymoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/repository/EventAuditRepository.kt:17]: should 'BOOKED_VISIT' be replaced with enum name?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/service/VisitService.kt [hunk 3]
```diff
@@ -66,10 +79,32 @@ class VisitService(
...
     // check if the existing visit is in the past
     validateVisitStartDate(visit, "changed")
-    return reserveVisitSlot(bookingReference, reserveVisitSlotDto)
+    val visitDto = createApplication(bookingReference, reserveVisitSlotDto)
+
+    saveEventAudit(
+      reserveVisitSlotDto.actionedBy,
+      visitDto,
+      if (visitDto.visitStatus == RESERVED) RESERVED_VISIT else CHANGING_VISIT,
+      NOT_KNOWN,
```
Comment (dhirajshettymoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/service/VisitService.kt:88]: will it be NOT_KNOWN for reserved visits and when the visit is booked we set the application type. is that right?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/service/VisitService.kt [hunk 4]
```diff
@@ -122,10 +157,11 @@ class VisitService(
...
     if (bookedVisit == null ||
-      (bookedVisit.prison.code != sessionTemplate.prisonCode) ||
-      (bookedVisit.prisonerId != reserveVisitSlotDto.prisonerId) ||
-      (bookedVisit.visitRestriction != reserveVisitSlotDto.visitRestriction) ||
-      (bookedVisit.visitStart.compareTo(reserveVisitSlotDto.startTimestamp) != 0)
+      bookedVisit.prison.code != sessionTemplate.prisonCode ||
+      bookedVisit.prisonerId != reserveVisitSlotDto.prisonerId ||
+      bookedVisit.visitRestriction != reserveVisitSlotDto.visitRestriction ||
+      bookedVisit.visitStart.compareTo(reserveVisitSlotDto.startTimestamp) != 0 ||
+      !bookedVisit.sessionTemplateReference.equals(reserveVisitSlotDto.sessionTemplateReference)
```
Comment (dhirajshettymoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/service/VisitService.kt:164]: is there no need for a date check as we discussed earlier?

File: src/main/resources/db/migration/V2_29__add_event_audit.sql [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+-- Copy data over from session_template table
+INSERT INTO event_audit (booking_reference, application_reference, session_template_reference, type, application_method_type, actioned_by, create_timestamp)
+    (SELECT reference,
+            application_reference,
+            session_template_reference,
+            CASE
+                WHEN visit_status   = 'RESERVED' THEN 'RESERVED_VISIT'
+                WHEN visit_status   = 'CHANGING' THEN 'CHANGING_VISIT'
+                WHEN visit_status   = 'BOOKED' THEN 'BOOKED_VISIT'
+                WHEN outcome_status = 'SUPERSEDED_CANCELLATION' THEN 'UPDATED_VISIT'
```
Comment (dhirajshettymoj) [inline src/main/resources/db/migration/V2_29__add_event_audit.sql:26]: will it have UPDATED_VISIT on the first entry?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/repository/EventAuditRepository.kt (hunk 2 above)
Comment (AledWEvans) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/repository/EventAuditRepository.kt:17]: I suppose si

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/service/MigrateVisitService.kt [hunk 6]
```diff
@@ -139,7 +155,17 @@ class MigrateVisitService(
...
-    visitEntity.cancelledBy = cancelVisitDto.actionedBy
+
+    eventAuditRepository.saveAndFlush(
+      EventAudit(
+        actionedBy = cancelVisitDto.actionedBy,
+        bookingReference = visitEntity.reference,
+        applicationReference = visitEntity.applicationReference,
+        sessionTemplateReference = visitEntity.sessionTemplateReference,
+        type = CANCELED_VISIT,
+        applicationMethodType = NOT_KNOWN,
```
[the remark this answers is not in the document]
Comment (AledWEvans) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/service/MigrateVisitService.kt:166]: requirements say other wise, we know if it's migrated by other means Legacy table.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/repository/EventAuditRepository.kt [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+      "ORDER BY id DESC LIMIT 1 ",
+    nativeQuery = true,
+  )
+  fun findLastEventByBookingReference(bookingReference: String): EventAudit
+
+  @Transactional
+  @Modifying
+  @Query(
+    "Update event_audit SET application_method_type=:#{#applicationMethodType.name()}  " +
+      "WHERE application_reference = :applicationReference AND type in ('RESERVED_VISIT','CHANGING_VISIT')",
```
[the remark this answers is not in the document]
Comment (AledWEvans) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/repository/EventAuditRepository.kt:27]: I think it's ugly though but is dynamic I suppose

File: src/main/resources/db/migration/V2_29__add_event_audit.sql (hunk 5 above)
Comment (AledWEvans) [reply src/main/resources/db/migration/V2_29__add_event_audit.sql:26]: don't think so,.. will ether be RESERVED_VISIT or CHANGING_VISIT

Comment (AledWEvans) [reply src/main/resources/db/migration/V2_29__add_event_audit.sql:26]: this is the code that needs most attention

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/service/VisitService.kt (hunk 4 above)
Comment (dhirajshettymoj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/service/VisitService.kt:164]: oh sorry I thought that wa time comparison.. you are right :)

File: src/main/resources/db/migration/V2_29__add_event_audit.sql (hunk 5 above)
Comment (dhirajshettymoj) [reply src/main/resources/db/migration/V2_29__add_event_audit.sql:26]: I mean if the visit is Updated will it end up with UPDATED_VISIT as the first entry?

**Review by dhirajshettymoj — DISMISSED** [review]
LGTM

**Review by dhirajshettymoj — APPROVED** [review]
LGTM

**Review by dhirajshettymoj — APPROVED** [review]
LGTM

**Pull request merged**